### ISSUE TITLE
[IDEA] Invisible route parameters

### ISSUE THREAD

**Issue by elderapo** [typehero/type-route#30]
No idea if "invisible route params" are a thing but let me explain...

Let's say I am creating a simple crypto exchange website. On the homepage (`/`), I want to have a minimalistic form (with default values) for selecting crypto type (BTC, BCH, ETH...), amount and just "BUY" button. Now after the user clicks "BUY" I'd like to redirect him to the real create order page (`/create-order`) which will have exact same form as on homepage + some other inputs (address, payment method...). Now the trick would be to somehow forward values from the homepage (currently selected crypto type, amount) to the create order page without using URL params.

If the user just types `/create-order` in an URL bar there won't be anything to forward which means that "invisible params" should be in fact optional. In this particular case default values should be used just as on the homepage.

This problem could be solved with optional query parameters but IMO it doesn't feel right to put everything in an URL bar. There are probably other use cases where something like this would very useful.

I guess implementation should be fairly easy and API could be just 2 new route params `invisible.param.string`, `invisible.param.number`.

Example:

```ts
export const router = createRouter({
  createOrder: defineRoute(
    {
      cryptoType: "invisible.param.string",
      amount: "invisible.param.number"
    },
    () => "/create-order"
  )
});

// ok
router.routes.createOrder.push({
  cryptoType: "ETH",
  amount: 123
});
// ok
router.routes.createOrder.push({
  amount: 123
});
// ok
router.routes.createOrder.push();
```

**Comment by bradenhs** [comment]
Thanks for posting this. The way we'd likely make this a possibility is via exposing a way to interact with the location state api. See this issue for a discussion on that: https://github.com/type-route/type-route/issues/29. This would actual be fairly easy to implement. Maybe instead of "invisible" we would just call these "state" params since that matches the browser api more closely. Regardless, thanks for posting this. I anticipate something like this will probably be added to Type Route soon.

**Comment by seivan** [comment]
Maybe I misunderstood, but isn't this essentially what optional params (supported today) is? 

> to the create order page without using URL params.

This isn't the routers responsibility as far as I can tell. 
If you want to seed a route (a component) with values, then you can pass them in as props
so your `/create-order` url would remain "naked". 

But you can also support query params, if you want to link to a ready made order form.

> means that "invisible params" should be in fact optional. In this particular case default values should be used just as on the homepage.

There are already optional params. 
If you want to see your URL with these "default" when visiting a "naked"`/create-order` then, you can do that with a replace under `listen( route => ... )` and replace the current route with your seed.

**Comment by elderapo** [comment]
> This isn't the routers responsibility as far as I can tell.
> If you want to seed a route (a component) with values, then you can pass them in as props
> so your `/create-order` url would remain "naked".

To do that you'd need a global state. Simple `router.routes.createRoute.push({ amount: 123, currency: "USD" })` from page at `/orders/:id` seems like a much better option.

> But you can also support query params, if you want to link to a ready made order form.

I don't like that. In that case, you'd need to update URL always when changing form values manually (user got redirected back to create order but decided to change amount) or you will end up with state mismatch (HTML form state values vs URL values). This doesn't feel right.

**Comment by seivan** [comment]
@elderapo just to make it clear.

You want to be able to go `/create-order` without having visible params? 
Then that should not be part of your router. 
If I see 
`router.routes.createRoute.push({ amount: 123, currency: "USD" })` in code, I am making the assumption that these are part of the URL. 

What you're asking for is just values in a component you are rendering, if that's a state derived from a hook or props passed is up to you, doesn't have to be a global state. 


> I don't like that. In that case, you'd need to update URL always when changing form values manually (user got redirected back to create order but decided to change amount) or you will end up with state mismatch (HTML form state values vs URL values). This doesn't feel right.

Not sure how you arrived at that conclusion. If you want to be able to link a ready made page you need to derive a state from whatever you begin from, if that's a URL then those are the values you derive your state from. 

In your case you want to be able to go to `/create-order` from a page and have params passed with it based on previous state without being visible in the URL (unless I've misunderstood) , this doesn't belong in the router. 

Use something like Mobx or Redux to manage that. 

Say you got a logged in page named `/dashboard`
You don't pass in your users `auth` state in the URL, it's "invisible", correct? That's similar notions. 

Your routes derive their state in a combination of your state and the params they get. 
You don't want params, that's fine, just derive from your state and keep it out of the routing. 

Messing up the API for invisible params just adds more confusion on how your URLS will look like and what's happening when a route is visited, not to mention that if you were to change your derived state, your URLS would break.

**Comment by elderapo** [comment]
@seivan Consider this scenario:

1. User visits website for the first time.
2. Goes to `/create-order`.
3. Form gets prefilled with currency `USD` and amount `123`.
4. User changes currency to `EUR` and amount to: `321`.
5. User click create order.
6. Order get created.
7. User gets redirected to `/order/:id`.
8. On this page user clicks repeat order.
9. Gets redirected to `/create-order`.
10. Now defaults for the form are currency `USD` and amount `123` but because of user action (edit order) form should get prefilled with values from order.

To achieve that with state params single `router.routes.createOrder.push({ currency: "EUR", amount:321 })` does the job. Without it I global state is required (or at least shared between these 2 routes) - that's what I am doing right now (using `mobx`) but it's not ergonomic.

This use case is not very unique. State params are even supported by browser history api.

**Comment by seivan** [comment]
Ah, yeah that makes more sense, thanks. 

Can it be more explicit at all portions if this is implemented? 
When defining the route, make it explicit if these sets of params are not visible in the url, while the others are.
When pushing, make it clear that these arguments are the non visible states being passed.

The only case where it doesn't matter, is probably when consuming it. 
`currentRoute.params` probably doesn't need a distinction.

**Comment by elderapo** [comment]
When defining a route and interpolating it only `path` params are available. `query` params are in URL as well but there there is not much place for customization (they're always at the end in the same format: `?first=val&second=val`). I think it makes sense to keep `type-route` API simple as it is and just add `state.param.(number/string)` option to params object:

```ts
defineRoute(
  {
    userId: "path.param.string",
    page: "query.param.number.optional",
    someStateParam: "state.param.number" // always optional because it's state param
  },
  p => `/users/${p.userId}/activity`
);
```

Information that this is a state param should be enough for the dev to know that param is not a part of the URL.

@bradenhs, have you thought about adding the possibility to specify the default values for optional params in route definition? Just a thought...

**Comment by bradenhs** [comment]
I have not thought about adding default values for optional params but really like the idea! I'm working on putting together some api updates for the defineRoute api now to support state params, parameters of types besides string/number, trailing path parameters among other things and have some thoughts about how default parameter values could work with that updated api. When its in a slightly more polished state I'll be sure to share to get some feedback.

**Comment by bradenhs** [comment]
This PR adds state parameters: https://github.com/type-route/type-route/pull/35. Any feedback on the new api would be appreciated.

**Comment by bradenhs** [comment]
Released support for this in `0.4.0`! Feel free to reopen if you feel that release doesn't address this fully.

**Comment by seivan** [comment]
Been using your fork for a while, and it's simply amazing. Thanks a lot dude! 
Also thank you @elderapo for being patient with me!
